src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Threading;
+
+namespace Npgsql.Util
+{
+    /// <summary>
+    /// Internal struct wrapping a <see cref="CancellationTokenSource"/> for timeouts.
+    /// </summary>
+    /// <remarks>
+    /// Since there's no way to reset a <see cref="CancellationTokenSource"/> once it was cancelled,
+    /// we need to make sure that an existing cancellation token source hasn't been cancelled,
+    /// every time we start it (see https://github.com/dotnet/runtime/issues/4694).
+    /// </remarks>
+    class TimeoutCancellationTokenSourceWrapper : IDisposable
```

Comment:
First think about this is to make it a struct.

Reply:
Not sure that's important (these get allocated only per physical connection etc.). IMHO it's OK either as a class or struct.

Reply:
I'd say that `CancellationTokenSource` is a class too, so we're allocating anyways.
The additional 8 bytes we allocate for the `Timeout` property will not result in any measurable GC pressure.
I'll leave it as a class.

Reply:
You allocate more than 8 bytes, because of the header prefixing each heap allocated object plus a pointer to the source. It also adds one more object with each allocation which will be analyzed by GC and reallocated.

Take a look at `ManualResetValueTaskSourceCore<TResult>`. It's a structure stored in the object owning it. You did the same helper which has only one owner for its lifetime.

Reply:
Ok, I still think that adding all the null checks isn't worthwhile.

Also see https://github.com/npgsql/npgsql/pull/3215#discussion_r499537774